### PR TITLE
Add Create BaremetalHost form to metal3 plugin

### PR DESCRIPTION
**Pull request by flofuchs**
This patch adds a new page to create new Baremetal hosts to the metal3 plugin.

### CODE REVIEW

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          setHasBaremetalhost(true);
+          setError(err);
+        });
+    }
+  }
+
+  const createBaremetalHost = async (secret, bmo) => {
+    const results = [];
+    const createFunc = hasBaremetalhost ? k8sUpdate : k8sCreate;
+    results.push(await k8sCreate(SecretModel, secret));
```
[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: The reasoning behind that was that if there is a newly discovered host there should not yet be a secret.

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  kind: BaremetalHostModel.kind,
+  metadata: {
+    name,
+    namespace,
+  },
+  spec: {
+    bmc: {
+      address: controller,
+      credentialsName: getSecretName(name),
+    },
```
[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: Done

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    name,
+  },
+  data: {
+    username: btoa(username),
+    password: btoa(password),
+  },
+  type: 'Opaque',
+});
+
+const getBaremetalHostObject = (name, namespace, controller) => ({
```
[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: Done

[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: Done

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import { ButtonBar, history, resourcePathFromModel } from '@console/internal/components/utils';
+import { getActiveNamespace } from '@console/internal/actions/ui';
+import { SecretModel } from '@console/internal/models';
+import { k8sCreate, k8sGet, k8sUpdate, referenceForModel } from '@console/internal/module/k8s';
+
+import { BaremetalHostModel } from '../models';
+
+const getSecretName = name => `${name}-secret`;
+
+const getSecret = (name, namespace, username, password) => ({
```
[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: Done

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    },
+  },
+});
+
+
+const AddHost: React.FC = ({ match }) => {
+
+  const title = 'Create Bare Metal Host';
+
+  const [formValue, setFormValue] = useState({});
```
[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: Done

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 6]
```diff
@@ -0,0 +1,188 @@
+import React, { useState } from 'react';
```
[the remark this answers is not in the document]
Comment (flofuchs) [reply frontend/packages/metal3-plugin/src/components/add-host.tsx]: Done

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                id="host-online"
+                label={`Host will be ${formValue.online ? 'online' : 'offline'} after creation.`}
+                isChecked={formValue.online}
+                onChange={setOnlineState}
+              />
+            </p>
+          </div>
+
+          <ButtonBar errorMessage={error} inProgress={inProgress}>
+            <button
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]:
```suggestion
            <Button
                type="submit"
                disabled={
                  !formValue.name ||
                  !formValue.controller ||
                  !formValue.username ||
                  !formValue.password
                }
                id="save-changes"
              >
                Create
              </Button>
```

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                !formValue.password
+              }
+              className="btn btn-primary"
+              id="save-changes"
+            >
+              Create
+            </button>
+            <Link
+              to={`/k8s/ns/${getActiveNamespace()}/baremetalhosts`}
+              className="btn btn-default"
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]: className="pf-c-button pf-m-secondary"

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+import { Switch } from '@patternfly/react-core';
+import { ButtonBar, history, resourcePathFromModel } from '@console/internal/components/utils';
+import { getActiveNamespace } from '@console/internal/actions/ui';
+import { SecretModel } from '@console/internal/models';
+import { k8sCreate, k8sGet, k8sUpdate } from '@console/internal/module/k8s';
+import { BaremetalHostModel } from '../models';
+
+interface AddHostFormFields {
+  name: string;
+  controller: string;
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]: I'd rename this to `managementAddress`

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+              defaultValue={formValue.name}
+              onChange={handleChange}
+              className="form-control"
+              required
+            />
+          </div>
+
+          <div className="form-group">
+            <label htmlFor="controller" className="control-label co-required">
+              Controller
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]: This should be called `Management Controller Address`

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    online,
+    bmc: {
+      address: controller,
+      credentialsName: getSecretName(name),
+    },
+    hardwareProfile: 'dell',
+  },
+});
+
+const AddHost: React.FC = ({ match }) => {
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]:
```suggestion
type MatchParams = { name: string };
type AddHostProps = RouteComponentProps<MatchParams>;

const AddHost: React.FC<AddHostProps> = ({ match }) => {
```

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 12]
```diff
@@ -63,16 +63,49 @@ const buildBaremetalHostObject = (name, namespace, controller, online) => ({
...
   const [error, setError] = React.useState('');
+  const [baremetalhost, setBaremetalhost] = React.useState({});
+  const [isFetching, setIsFetching] = React.useState(false);
+  const [hasBaremetalhost, setHasBaremetalhost] = React.useState(false);
+
+  // Get the name from the URL (if there is one)
+  const { params: { name: nameFromUrl } = {} } = match;
+  // If the name is set in the URL we'd like to set it in the form value
+  // and retrieve the object from the API.
+  if (nameFromUrl) {
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]: This is a side effect and needs to be wrapped in useEffect hook. We don't want to do it on every render

File: frontend/packages/metal3-plugin/src/components/add-host.tsx [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+              className="form-control"
+              required
+            />
+          </div>
+
+          <div className="formGroup">
+            <p>
+              <Switch
+                id="host-online"
+                label={`Host will be ${formValue.online ? 'online' : 'offline'} after creation.`}
```
Comment (jtomasek) [inline frontend/packages/metal3-plugin/src/components/add-host.tsx]: Making the label dynamic may be slightly confusing, maybe a simple `Make host online after creation` could be better?

**Pull request merged**